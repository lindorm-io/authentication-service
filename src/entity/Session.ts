import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";
import { SessionEvent } from "../enum";

export interface ISessionAgent {
  browser: string;
  geoIp: TObject<any>;
  os: string;
  platform: string;
  source: string;
  version: string;
}

export interface ISessionAuthorization {
  codeChallenge: string;
  codeMethod: string;
  deviceChallenge: string;
  email: string;
  id: string;
  otpCode: string;
  redirectUri: string;
  responseType: string;
}

export interface ISession extends IEntity {
  accountId: string;
  agent: ISessionAgent;
  authenticated: boolean;
  authorization: ISessionAuthorization;
  clientId: string;
  deviceId: string;
  expires: Date;
  grantType: string;
  refreshId: string;
  scope: string;
}

export interface ISessionOptions extends IEntityBaseOptions {
  accountId?: string;
  agent?: {
    browser?: string;
    geoIp?: TObject<any>;
    os?: string;
    platform?: string;
    source?: string;
    version?: string;
  };
  authenticated?: boolean;
  authorization: {
    codeChallenge: string;
    codeMethod: string;
    deviceChallenge?: string;
    email: string;
    id: string;
    otpCode?: string;
    redirectUri: string;
    responseType: string;
  };
  clientId: string;
  deviceId?: string;
  expires: Date;
  grantType: string;
  refreshId?: string;
  scope: string;
}

export class Session extends EntityBase implements ISession {
  private _accountId: string;
  private _agent: ISessionAgent;
  private _authenticated: boolean;
  private _authorization: ISessionAuthorization;
  private _clientId: string;
  private _deviceId: string;
  private _expires: Date;
  private _grantType: string;
  private _refreshId: string;
  private _scope: string;

  constructor(options: ISessionOptions) {
    super(options);

    this._accountId = options.accountId || null;
    this._agent = {
      browser: options.agent?.browser || null,
      geoIp: options.agent?.geoIp || null,
      os: options.agent?.os || null,
      platform: options.agent?.platform || null,
      source: options.agent?.source || null,
      version: options.agent?.version || null,
    };
    this._authenticated = options.authenticated || false;
    this._authorization = {
      codeChallenge: options.authorization.codeChallenge,
      codeMethod: options.authorization.codeMethod,
      deviceChallenge: options.authorization.deviceChallenge || null,
      email: options.authorization.email,
      id: options.authorization.id,
      otpCode: options.authorization.otpCode || null,
      redirectUri: options.authorization.redirectUri,
      responseType: options.authorization.responseType,
    };
    this._clientId = options.clientId;
    this._deviceId = options.deviceId || null;
    this._expires = options.expires;
    this._grantType = options.grantType;
    this._refreshId = options.refreshId || null;
    this._scope = options.scope;
  }

  public get accountId(): string {
    return this._accountId;
  }
  public set accountId(accountId: string) {
    this._accountId = accountId;
    this.addEvent(SessionEvent.ACCOUNT_ID_CHANGED, { accountId: this._accountId });
  }

  public get agent(): ISessionAgent {
    return this._agent;
  }

  public get authenticated(): boolean {
    return this._authenticated;
  }
  public set authenticated(authenticated: boolean) {
    this._authenticated = authenticated;
    this.addEvent(SessionEvent.AUTHENTICATED_CHANGED, { authenticated: this._authenticated });
  }

  public get authorization(): ISessionAuthorization {
    return this._authorization;
  }

  public get clientId(): string {
    return this._clientId;
  }

  public get deviceId(): string {
    return this._deviceId;
  }

  public get expires(): Date {
    return this._expires;
  }
  public set expires(expires: Date) {
    this._expires = expires;
    this.addEvent(SessionEvent.EXPIRES_CHANGED, { expires: this._expires });
  }

  public get grantType(): string {
    return this._grantType;
  }

  public get refreshId(): string {
    return this._refreshId;
  }
  public set refreshId(refreshId: string) {
    this._refreshId = refreshId;
    this.addEvent(SessionEvent.REFRESH_ID_CHANGED, { refreshId: this._refreshId });
  }

  public get scope(): string {
    return this._scope;
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== SessionEvent.CREATED) continue;
      throw new EntityCreationError("Session");
    }

    this.addEvent(SessionEvent.CREATED, {
      accountId: this._accountId,
      agent: this._agent,
      authenticated: this._authenticated,
      authorization: this._authorization,
      clientId: this._clientId,
      deviceId: this._deviceId,
      expires: this._expires,
      grantType: this._grantType,
      refreshId: this._refreshId,
      scope: this._scope,
    });
  }
}
