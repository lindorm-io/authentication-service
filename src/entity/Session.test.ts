import MockDate from "mockdate";
import { Session } from "./Session";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");
const date = new Date("2020-01-01 08:00:00.000");

describe("Session.ts", () => {
  let session: Session;

  beforeEach(() => {
    session = new Session({
      accountId: "accountId",
      agent: {
        browser: "browser",
        geoIp: { geoIp: "geoIp" },
        os: "os",
        platform: "platform",
        source: "source",
        version: "version",
      },
      authenticated: true,
      authorization: {
        codeChallenge: "codeChallenge",
        codeMethod: "codeMethod",
        deviceChallenge: "deviceChallenge",
        email: "email@lindorm.io",
        id: "id",
        otpCode: "otpCode",
        redirectUri: "redirectUri",
        responseType: "responseType",
      },
      clientId: "clientId",
      deviceId: "deviceId",
      expires: date,
      grantType: "grantType",
      refreshId: "refreshId",
      scope: ["scope"],
    });
  });

  test("should have all data", () => {
    expect(session).toMatchSnapshot();
  });

  test("should have optional data", () => {
    session = new Session({
      authorization: {
        codeChallenge: "codeChallenge",
        codeMethod: "codeMethod",
        email: "email@lindorm.io",
        id: "id",
        redirectUri: "redirectUri",
        responseType: "responseType",
      },
      clientId: "clientId",
      expires: date,
      grantType: "grantType",
      scope: ["scope"],
    });

    expect(session).toMatchSnapshot();
  });

  test("should create", () => {
    session.create();
    expect(session.events).toMatchSnapshot();
  });

  test("should get/set accountId", () => {
    expect(session.accountId).toMatchSnapshot();

    session.accountId = "new-accountId";

    expect(session.accountId).toMatchSnapshot();
    expect(session.events).toMatchSnapshot();
  });

  test("should get agent", () => {
    expect(session.agent).toMatchSnapshot();
  });

  test("should get/set authenticated", () => {
    expect(session.authenticated).toMatchSnapshot();

    session.authenticated = false;

    expect(session.authenticated).toMatchSnapshot();
    expect(session.events).toMatchSnapshot();
  });

  test("should get authorization", () => {
    expect(session.authorization).toMatchSnapshot();
  });

  test("should get clientId", () => {
    expect(session.clientId).toMatchSnapshot();
  });

  test("should get/set expires", () => {
    expect(session.expires).toMatchSnapshot();

    session.expires = new Date("2021-01-01 00:00:01");

    expect(session.expires).toMatchSnapshot();
    expect(session.events).toMatchSnapshot();
  });

  test("should get grantType", () => {
    expect(session.grantType).toMatchSnapshot();
  });

  test("should get/set refreshId", () => {
    expect(session.refreshId).toMatchSnapshot();

    session.refreshId = "new-refreshId";

    expect(session.refreshId).toMatchSnapshot();
    expect(session.events).toMatchSnapshot();
  });

  test("should get scope", () => {
    expect(session.scope).toMatchSnapshot();
  });
});
