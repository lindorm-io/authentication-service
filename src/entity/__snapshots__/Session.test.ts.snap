// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session.ts should create 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "session_created",
    "payload": Object {
      "accountId": "accountId",
      "agent": Object {
        "browser": "browser",
        "geoIp": Object {
          "geoIp": "geoIp",
        },
        "os": "os",
        "platform": "platform",
        "source": "source",
        "version": "version",
      },
      "authenticated": true,
      "authorization": Object {
        "codeChallenge": "codeChallenge",
        "codeMethod": "codeMethod",
        "deviceChallenge": "deviceChallenge",
        "email": "email@lindorm.io",
        "id": "id",
        "otpCode": "otpCode",
        "redirectUri": "redirectUri",
        "responseType": "responseType",
      },
      "clientId": "clientId",
      "deviceId": "deviceId",
      "expires": 2020-01-01T07:00:00.000Z,
      "grantType": "grantType",
      "refreshId": "refreshId",
      "scope": Array [
        "scope",
      ],
    },
  },
]
`;

exports[`Session.ts should get agent 1`] = `
Object {
  "browser": "browser",
  "geoIp": Object {
    "geoIp": "geoIp",
  },
  "os": "os",
  "platform": "platform",
  "source": "source",
  "version": "version",
}
`;

exports[`Session.ts should get authorization 1`] = `
Object {
  "codeChallenge": "codeChallenge",
  "codeMethod": "codeMethod",
  "deviceChallenge": "deviceChallenge",
  "email": "email@lindorm.io",
  "id": "id",
  "otpCode": "otpCode",
  "redirectUri": "redirectUri",
  "responseType": "responseType",
}
`;

exports[`Session.ts should get clientId 1`] = `"clientId"`;

exports[`Session.ts should get grantType 1`] = `"grantType"`;

exports[`Session.ts should get scope 1`] = `
Array [
  "scope",
]
`;

exports[`Session.ts should get/set accountId 1`] = `"accountId"`;

exports[`Session.ts should get/set accountId 2`] = `"new-accountId"`;

exports[`Session.ts should get/set accountId 3`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "session_account_id_changed",
    "payload": Object {
      "accountId": "new-accountId",
    },
  },
]
`;

exports[`Session.ts should get/set authenticated 1`] = `true`;

exports[`Session.ts should get/set authenticated 2`] = `false`;

exports[`Session.ts should get/set authenticated 3`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "session_authenticated_changed",
    "payload": Object {
      "authenticated": false,
    },
  },
]
`;

exports[`Session.ts should get/set expires 1`] = `2020-01-01T07:00:00.000Z`;

exports[`Session.ts should get/set expires 2`] = `2020-12-31T23:00:01.000Z`;

exports[`Session.ts should get/set expires 3`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "session_expires_changed",
    "payload": Object {
      "expires": 2020-12-31T23:00:01.000Z,
    },
  },
]
`;

exports[`Session.ts should get/set refreshId 1`] = `"refreshId"`;

exports[`Session.ts should get/set refreshId 2`] = `"new-refreshId"`;

exports[`Session.ts should get/set refreshId 3`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "session_refresh_id_changed",
    "payload": Object {
      "refreshId": "new-refreshId",
    },
  },
]
`;

exports[`Session.ts should have all data 1`] = `
Session {
  "_accountId": "accountId",
  "_agent": Object {
    "browser": "browser",
    "geoIp": Object {
      "geoIp": "geoIp",
    },
    "os": "os",
    "platform": "platform",
    "source": "source",
    "version": "version",
  },
  "_authenticated": true,
  "_authorization": Object {
    "codeChallenge": "codeChallenge",
    "codeMethod": "codeMethod",
    "deviceChallenge": "deviceChallenge",
    "email": "email@lindorm.io",
    "id": "id",
    "otpCode": "otpCode",
    "redirectUri": "redirectUri",
    "responseType": "responseType",
  },
  "_clientId": "clientId",
  "_created": 2020-01-01T07:00:00.000Z,
  "_deviceId": "deviceId",
  "_events": Array [],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_grantType": "grantType",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_refreshId": "refreshId",
  "_scope": Array [
    "scope",
  ],
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`Session.ts should have optional data 1`] = `
Session {
  "_accountId": null,
  "_agent": Object {
    "browser": null,
    "geoIp": null,
    "os": null,
    "platform": null,
    "source": null,
    "version": null,
  },
  "_authenticated": false,
  "_authorization": Object {
    "codeChallenge": "codeChallenge",
    "codeMethod": "codeMethod",
    "deviceChallenge": null,
    "email": "email@lindorm.io",
    "id": "id",
    "otpCode": null,
    "redirectUri": "redirectUri",
    "responseType": "responseType",
  },
  "_clientId": "clientId",
  "_created": 2020-01-01T07:00:00.000Z,
  "_deviceId": null,
  "_events": Array [],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_grantType": "grantType",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_refreshId": null,
  "_scope": Array [
    "scope",
  ],
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
